src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/GroupCollection.cs

```diff
@@ -205,7 +205,55 @@ void IList.RemoveAt(int index)
             set { throw new NotSupportedException(SR.NotSupported_ReadOnlyCollection); }
         }
 
-        private sealed class Enumerator : IEnumerator<Group>
+        IEnumerator<KeyValuePair<string, Group>> IEnumerable<KeyValuePair<string, Group>>.GetEnumerator()
+        {
+            return new Enumerator(this);
+        }
+
+        public bool TryGetValue(string key, out Group value)
+        {
+            var group = this[key];
+            if (group == Group.s_emptyGroup)
+            {
+                value = null;
+                return false;
+            }
+            value = group;
+            return true;
+        }
+
+        public bool ContainsKey(string key)
+        {
+            return _match._regex.GroupNumberFromName(key) >= 0;
+        }
+
+        public IEnumerable<string> Keys
+        {
+            get
+            {
+                var enumerator = new Enumerator(this);
```

Comment:
@stephentoub does that seem right to you? Now we are allocating because of the enumerator instantiation, which is odd to me.

Reply:
ping @stephentoub

Reply:
> Yeah but if you don't want to risk exposing something internally the enumerator allocation seemed the cheapest alternative

Right, either you expose the internal data structure or allocate a wrapper.  It's up to you, @ViktorHofer.

Reply:
I'm fine with the allocation. This is absolutely not a hot path.